runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java

```diff
@@ -4058,6 +4067,106 @@ public void testStuckCommit() throws Exception {
         removeDynamicFields(result.get(1L)));
   }
 
+  @Test
+  public void testSwitchStreamingWorkerHarness() throws Exception {
+    if (!streamingEngine) {
+      return;
+    }
+
+    List<ParallelInstruction> instructions =
+        Arrays.asList(
+            makeSourceInstruction(StringUtf8Coder.of()),
+            makeSinkInstruction(StringUtf8Coder.of(), 0));
+
+    // Start with CloudPath.
+    DataflowWorkerHarnessOptions options =
+        createTestingPipelineOptions("--isWindmillServiceDirectPathEnabled=false");
+
+    StreamingDataflowWorker worker =
+        makeWorker(
+            defaultWorkerParams()
+                .setOptions(options)
+                .setInstructions(instructions)
+                .publishCounters()
+                .build());
+
+    GrpcDispatcherClient mockDispatcherClient = mock(GrpcDispatcherClient.class);
+
+    // FanOutStreamingEngineWorkerHarness creates
+    // CloudWindmillMetadataServiceV1Alpha1Stub and expects the stream to
+    // successfully start. Mocking it here.
+    Channel mockChannel = mock(Channel.class);
+    ClientCall<WorkerMetadataRequest, WorkerMetadataResponse> mockClientCall =
+        mock(ClientCall.class);
+    when(mockChannel.newCall(
+            eq(CloudWindmillMetadataServiceV1Alpha1Grpc.getGetWorkerMetadataMethod()), any()))
+        .thenReturn(mockClientCall);
+    when(mockDispatcherClient.getWindmillMetadataServiceStubBlocking())
+        .thenReturn(CloudWindmillMetadataServiceV1Alpha1Grpc.newStub(mockChannel));
+    java.lang.reflect.Field dispatcherClientField =
+        StreamingDataflowWorker.class.getDeclaredField("dispatcherClient");
+    dispatcherClientField.setAccessible(true);
+    dispatcherClientField.set(worker, mockDispatcherClient);
```

Comment:
Prefer fakes/mocks over using reflection directly.

Here we can update FakeWindmillServer.java to setup fake metadata responses.